kotsadm/pkg/kurl/delete_node.go

```diff
@@ -0,0 +1,227 @@
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	configutil "k8s.io/kubernetes/cmd/kubeadm/app/util/config"
+)
+
+func DeleteNode(ctx context.Context, client kubernetes.Interface, restconfig *rest.Config, node *corev1.Node) error {
+	err := client.CoreV1().Nodes().Delete(ctx, node.Name, metav1.DeleteOptions{})
+	if err != nil {
+		return errors.Wrap(err, "delete node")
+	}
+
+	err = purgeOSD(ctx, client, restconfig, node)
+	if err != nil {
+		return errors.Wrap(err, "purge OSD on node")
+	}
+
+	_, isPrimary := node.Labels["node-role.kubernetes.io/master"]
+	if isPrimary {
+		logger.Infof("Node %s is a primary: running etcd and kubeadm endpoints purge steps", node.Name)
+		if err := purgePrimary(ctx, client, node); err != nil {
+			return err
+		}
+	} else {
+		logger.Debugf("Node %s is not a primary: skipping etcd and kubeadm endpoints purge steps", node.Name)
+	}
+	return nil
+}
+
+func purgeOSD(ctx context.Context, client kubernetes.Interface, restconfig *rest.Config, node *corev1.Node) error {
+	// 1. Find the Deployment for the OSD on the purged node and lookup its osd ID from labels
+	// before deleting it.
+	opts := metav1.ListOptions{
+		LabelSelector: "app=rook-ceph-osd",
+	}
+	osdDeployments, err := client.AppsV1().Deployments("rook-ceph").List(ctx, opts)
+
+	var osdID string
+
+	for _, deployment := range osdDeployments.Items {
+		hostname := deployment.Spec.Template.Spec.NodeSelector["kubernetes.io/hostname"]
+
+		if hostname == node.Name {
+			osdID = deployment.Labels["ceph-osd-id"]
+			logger.Infof("Deleting OSD deployment on node %s", node.Name)
+			if err := client.AppsV1().Deployments("rook-ceph").Delete(ctx, deployment.Name, metav1.DeleteOptions{}); err != nil {
+				return errors.Wrapf(err, "delete deployment %s", deployment.Name)
+			}
+			break
+		}
+	}
+
+	if osdID == "" {
+		logger.Infof("Failed to find ceph osd id for node %s", node.Name)
+		// Not an error - rook probably isn't in use
+		return nil
+	}
+
+	logger.Infof("Purging ceph OSD with id %s", osdID)
+
+	// 2. Using the osd ID discovered in step 1, exec into the Rook operator pod and run the ceph
+	// command to purge the OSD
+	rookOperatorLabels := map[string]string{"app": "rook-ceph-operator"}
+	listOpts := metav1.ListOptions{
+		LabelSelector: labels.SelectorFromSet(rookOperatorLabels).String(),
+	}
+	pods, err := client.CoreV1().Pods("rook-ceph").List(ctx, listOpts)
+	if err != nil {
+		return errors.Wrap(err, "list Rook Operator pods")
+	}
+	if len(pods.Items) == 0 {
+		return fmt.Errorf("Failed to purge OSD: found 0 Rook Operator pods")
+	}
+	// ignore error
+	SyncExec(client.CoreV1(), restconfig, "rook-ceph", pods.Items[0].Name, "rook-ceph-operator", "ceph", "osd", "down", osdID)
```

Comment:
why ignore error for `ceph osd down`?

Reply:
I'm no longer sure there is a benefit to marking the osd down before purgng. I'm going to remove this line.